File: java/src/test/java/jsontag/SerializeTest.java
package jsontag;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class SerializeTest {

    private final ObjectMapper mapper = new ObjectMapper();

    // Interop JSON — identical strings used across all 12 language libraries
    static final String CIRCLE_JSON = """
            {"#type":"Circle","radius":4}""";
    static final String RECTANGLE_JSON = """
            {"#type":"Rectangle","width":10,"height":5}""";
    static final String CAR_JSON = """
            {"#type":"Car","make":"Toyota"}""";
    static final String TRUCK_JSON = """
            {"#type":"Truck","payloadTons":5}""";

    // Jackson emits .0 for whole-number doubles
    static final String JACKSON_CIRCLE_JSON = """
            {"#type":"Circle","radius":4.0}""";
    static final String JACKSON_RECTANGLE_JSON = """
            {"#type":"Rectangle","width":10.0,"height":5.0}""";
    static final String JACKSON_TRUCK_JSON = """
            {"#type":"Truck","payloadTons":5.0}""";

    @Test
    void serialize_whenSerializingCircle_matchesInteropFormat() throws Exception {
        // Act
        String json = mapper.writeValueAsString(new Circle(4.0));

        // Assert
        assertEquals(JACKSON_CIRCLE_JSON, json);
    }

    @Test
    void serialize_whenSerializingRectangle_matchesInteropFormat() throws Exception {
        // Act
        String json = mapper.writeValueAsString(new Rectangle(10.0, 5.0));

        // Assert
        assertEquals(JACKSON_RECTANGLE_JSON, json);
    }

    @Test
    void serialize_whenSerializingCar_matchesInteropFormat() throws Exception {
        // Act
        String json = mapper.writeValueAsString(new PersonalVehicle("Toyota"));

        // Assert
        assertEquals(CAR_JSON, json);
    }

    @Test
    void serialize_whenSerializingTruck_matchesInteropFormat() throws Exception {
        // Act
        String json = mapper.writeValueAsString(new TransportationVehicle(5.0));

        // Assert
        assertEquals(JACKSON_TRUCK_JSON, json);
    }
}

File: kotlin/src/test/kotlin/jsontag/SerializeTest.kt
package jsontag

import kotlin.test.Test
import kotlin.test.assertEquals

class SerializeTest {

    // Interop JSON — identical strings used across all 12 language libraries
    companion object {
        const val CIRCLE_JSON = """{"#type":"Circle","radius":4}"""
        const val RECTANGLE_JSON = """{"#type":"Rectangle","width":10,"height":5}"""
        const val CAR_JSON = """{"#type":"Car","make":"Toyota"}"""
        const val TRUCK_JSON = """{"#type":"Truck","payloadTons":5}"""

        // kotlinx.serialization emits .0 for whole-number Doubles
        const val KOTLINX_CIRCLE_JSON = """{"#type":"Circle","radius":4.0}"""
        const val KOTLINX_RECTANGLE_JSON = """{"#type":"Rectangle","width":10.0,"height":5.0}"""
        const val KOTLINX_TRUCK_JSON = """{"#type":"Truck","payloadTons":5.0}"""
    }

    @Test
    fun `serialize, when serializing Circle, matches interop format`() {
        // Act
        val json = JsonTag.format.encodeToString(Shape.serializer(), Circle(radius = 4.0))

        // Assert
        assertEquals(KOTLINX_CIRCLE_JSON, json)
    }

    @Test
    fun `serialize, when serializing Rectangle, matches interop format`() {
        // Act
        val json = JsonTag.format.encodeToString(Shape.serializer(), Rectangle(width = 10.0, height = 5.0))

        // Assert
        assertEquals(KOTLINX_RECTANGLE_JSON, json)
    }

    @Test
    fun `serialize, when serializing Car, matches interop format`() {
        // Act
        val json = JsonTag.format.encodeToString(Vehicle.serializer(), PersonalVehicle(make = "Toyota"))

        // Assert
        assertEquals(CAR_JSON, json)
    }

    @Test
    fun `serialize, when serializing Truck, matches interop format`() {
        // Act
        val json = JsonTag.format.encodeToString(Vehicle.serializer(), TransportationVehicle(payloadTons = 5.0))

        // Assert
        assertEquals(KOTLINX_TRUCK_JSON, json)
    }
}
